File: leetcode/trie/explore/basic_operations/basic_operations.py
import collections

class TrieNode:
    def __init__(self):
        self.children = collections.defaultdict(TrieNode)
        self.isWord = False

class Trie:

    def __init__(self):
        self.root = TrieNode()
        

    def insert(self, word: str) -> None:
        curNode = self.root
        for letter in word:
            curNode = curNode.children[letter]
        curNode.isWord = True
        

    def search(self, word: str) -> bool:
        curNode = self.root
        for letter in word:
            curNode = curNode.children.get(letter)
            if curNode is None:
                return False

        return curNode.isWord

        

    def startsWith(self, prefix: str) -> bool:
        curNode = self.root
        for letter in prefix:
            curNode = curNode.children.get(letter)
            if curNode is None:
                return False
        return True
        


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)

File: leetcode/trie/explore/basic_operations/basic_operations.cpp
#include <unordered_map>

using namespace std;

struct TrieNode {
    unordered_map<char, TrieNode*> children;
    bool isEnd;

    TrieNode():isEnd(false) {}
};

class Trie {
private:
    TrieNode* root;
public:
    Trie() {
        root = new TrieNode();
    }
    
    /** Inserts a word into the trie. */
    void insert(string word) {
        TrieNode* current = this->root;
        for (auto c : word)
        {
            if (current->children.find(c) == current->children.end())
                current->children[c] = new TrieNode();
            current = current->children[c];
        }
        current->isEnd = true;
    }
    
    /** Returns if the word is in the trie. */
    bool search(string word) {
        TrieNode* current = this->root;
        for (auto c : word)
        {
            if (current->children.find(c) != current->children.end())
                current = current->children[c];
            else
                return false;
        }
        return current->isEnd;
    }
    
    /** Returns if there is any word in the trie that starts with the given prefix. */
    bool startsWith(string prefix) {
        TrieNode* current = this->root;
        for (auto c : prefix)
        {
            if (current->children.find(c) != current->children.end())
                current = current->children[c];
            else
                return false;
        }
        return true;
    }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */